protocols/abci/addressing-1174.md

```diff
@@ -0,0 +1,203 @@
+# Allowing Non-Determinism in `ProcessProposal` (a.k.a. addressing #1174)
+
+## Context
+
+### Consensus Properties
+
+Byzantine Fault Tolerant (BFT) Consensus is usually specified by the following invariants:
+
+- _agreement_: no two correct processes decide differently.
+- _validity_: the decided block `v` satisfies the predefined predicate `valid(v)`.
+- _termination_: all correct processes eventually decide.
+
+The consensus algorithm implemented in CometBFT (Tendermint) guarantees these invariants for each height.
+
+### ABCI interface
+
+In the version of ABCI (`v0.17.0`) that existed before ABCI 1.0 and 2.0 (a.k.a. ABCI++),
+the implementation of the `valid(v)` predicate was totally internal to CometBFT.
+The application had no direct say on the validity of a block,
+although it could (and still can) influence it indirectly via the best-effort ABCI call `CheckTx`.
+
+With the evolution of ABCI to ABCI 1.0 and 2.0, the `valid(v)` predicate now has two components:
+
+- the validity checks performed directly by CometBFT on blocks
+  (block format, hashes, etc; the same as in ABCI `v0.17.0`)
+- the validity checks that now the application can perform as part of `ProcessProposal`
+
+With these new validity checks:
+
+- consensus _agreement_ is not affected since the new structure of `valid(v)` in ABCI 1.0
+  does not accept any block `v` that would have been rejected in ABCI 0.17.0.
+  The new structure of `valid(v)` is thus more restrictive for any block `v`.
+- consensus _validity_ is not affected since we are changing the
+  structure of the `valid(v)` predicate, on which consensus _validity_ directly depends.
```

Comment:
I see that as the reason why _agreement_ is not affected.

Validity invariant is not affected because we are changing the internal "logic" of `valid(v)"... which is, say, a "parameter" of the invariant

Reply:
maybe we should discuss this synchronously but I read "we are changing the "internal logic" of `valid(v)" the same as "change in the structure of the `valid(v)` predicate only makes it more restrictive", that is, the reason why _validity_ is not affected.